mantidproject/mantid
# DirectILL: Option for detector grouping before Flat bkg calculation

The current implementation of the flat background calculation in `DirectILLCollectData` was found unsatisfactory by the ILL scientific users, as it almost always returned background equal to 0. This is believed to be caused by limited statistics of the measurement combined with the estimation assumption, that takes the lowest value of moving average (or fitting) as the value for the flat background.

This PR, extracted from the PR  #31597 to speed up delivery to users, addresses this issue by allowing the user to choose averaging window parameters.  It is now possible to group detectors horizontally (at the same vertical position and between different tubes) and vertically (in the same tube) to increase the statistics for the estimate of the flat background. The result was checked with ILL scientific users.

To test:

Ensure that you have UnitTest/ILL/PANTHER in your datasearch directories.
1. Run DirectILLCollectData on a PANTHER measurement, e.g. using the script below:

```
ei = CreateSingleValuedWorkspace(DataValue=19)
run = '002687.nxs'
DirectILLCollectData(
    Run=run,
    OutputWorkspace='test',
    IncidentEnergyCalibration='Energy Calibration ON',
    IncidentEnergyWorkspace=ei,
    FlatBkg='Flat Bkg ON',
    OutputFlatBkgWorkspace='flat',
)
```

Observe that the workspace named 'flat' contains only sporadic non-zero values in the current form.
Define two more properties of the DirectILLCollectData:

```
GroupDetHorizontallyBy=8,
GroupDetVerticallyBy=16
```

Observe how does the 'flat' workspace containing flat background behave, it should not contain zero background value anymore.


*There is no associated issue.*

Release notes and coumentation updated.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
